### PR TITLE
feat: validate from address on both worker and backend SES regions

### PR DESCRIPTION
**Pull request by tan-zx**
## Problem

Closes https://github.com/opengovsg/postmangovsg/issues/759

See [here](https://github.com/opengovsg/postmangovsg/issues/759#issuecomment-1257414111) for context

### CODE REVIEW

**Comment by tan-zx** [comment]
@stanleynguyen have made a related change to do the same checks for API emails (not sure if there is any reason against including these hmm might slow down speed of API calls? maybe don't need to check for every single call? wdyt hmm)

**Comment by stanleynguyen** [comment]
> @stanleynguyen have made a related change to do the same checks for API emails (not sure if there is any reason against including these hmm might slow down speed of API calls? maybe don't need to check for every single call? wdyt hmm)

It's quite unnecessary to do that on every call because they will get an error immediately on every call anyway, please remove it. We should formulate the verification as a separate endpoint for the API users to check if they wanna

**Comment by tan-zx** [comment]
> It's quite unnecessary to do that on every call because they will get an error immediately on every call anyway, please remove it. We should formulate the verification as a separate endpoint for the API users to check if they wanna

OK have removed. In the case where they get the error on each call, will the error message be informative though? 🤔

File: backend/src/email/services/custom-domain.service.ts [hunk 1]
```diff
@@ -50,7 +52,11 @@ const verifyCnames = async (
  * @returns DKIM tokens generated for the email address
  * @throws Error if the domain's dns do not have the cname records
  */
-const verifyEmailWithAWS = async (email: string): Promise<Array<string>> => {
+const verifyEmailWithAWS = async (
+  email: string,
+  ses: AWS.SES,
```
Comment (stanleynguyen) [inline backend/src/email/services/custom-domain.service.ts:57]: Is this unused?

Comment (stanleynguyen) [reply backend/src/email/services/custom-domain.service.ts:57]: Oh moved from global to local

File: backend/src/email/services/custom-domain.service.ts [hunk 2]
```diff
@@ -143,9 +153,18 @@ const storeFromAddress = async (
...
+  const dkimTokensBackend = await verifyEmailWithAWS(
+    email,
+    backendSes,
+    backendSesRegion
+  )
+  const dkimTokensWorker = await verifyEmailWithAWS(
+    email,
+    workerSes,
+    workerSesRegion
+  )
```
Comment (stanleynguyen) [inline backend/src/email/services/custom-domain.service.ts:165]: Rewrite this to 
```ts
await Promise.all([
  verifyEmailWithBackend,
  verifyEmailWithWorker,
])
```
to make more efficient use of the event loop cycle

Comment (tan-zx) [reply backend/src/email/services/custom-domain.service.ts:165]: Ohh good catch. Done.

**Comment by stanleynguyen** [comment]
> > It's quite unnecessary to do that on every call because they will get an error immediately on every call anyway, please remove it. We should formulate the verification as a separate endpoint for the API users to check if they wanna
> 
> OK have removed. In the case where they get the error on each call, will the error message be informative though? 🤔

They will get an internal server error, but we have logs of what happened so should be fine. I don't think this will be a problem cos ppl will read our guide first before integrating

File: backend/src/email/services/custom-domain.service.ts [hunk 3]
```diff
@@ -143,9 +153,11 @@ const storeFromAddress = async (
...
  */
 const verifyFromAddress = async (email: string): Promise<void> => {
-  const dkimTokens = await verifyEmailWithAWS(email)
-
-  await verifyCnames(dkimTokens, email)
+  const [dkimTokensBackend, dkimTokensWorker] = await Promise.all([
+    verifyEmailWithAWS(email, backendSes, backendSesRegion),
+    verifyEmailWithAWS(email, workerSes, workerSesRegion),
+  ])
+  await verifyCnames([...dkimTokensBackend, ...dkimTokensWorker], email)
```
Comment (stanleynguyen) [inline backend/src/email/services/custom-domain.service.ts:160]: We can refactor the `verifyCnames` function to use `Promise.all` as well. Do u wanna do it in this PR?

Comment (tan-zx) [reply backend/src/email/services/custom-domain.service.ts:162]: good catch, done (not 100% sure, but think it's right)

**Review by stanleynguyen — APPROVED** [review]

**Pull request merged**